### ISSUE TITLE
hasPath is not safe

### ISSUE THREAD

**Issue by char0n** [ramda/ramda#2738]
Would it be possible to make `hasPath` function safe ? It works as a predicate and I'd expect the predicates to be safe.

```js
hasPath(['a', 'b'], null); // throws Error
```
In [ramda-adjunct](https://github.com/char0n/ramda-adjunct/blob/master/src/hasPath.js) we have this function for more that a year, although is has recursive implementation it does the same but it's safe. The tests doesn't reflect what happens if the 2nd argument is not an object: https://github.com/ramda/ramda/blob/master/test/hasPath.js

**Comment by CrossEye** [comment]
Yes, I think we could do that.  This function answers the question, "Does this object have a value at the location represented by that path?"  I think we have a definitive answer of "no" (`false`) if the object doesn't exist.  You're right, too, that a predicate really shouldn't throw an error, if at all possible.

Would you be interested in creating a PR?

**Comment by char0n** [comment]
This also affected the behavior of `R.has` which is now also not safe because it uses `R.hasPath` under the hood.

**Comment by char0n** [comment]
Attaching the git patch. Will provide a proper PR in a couple of hours:
```
From fbd008e03000447b6e39ff6282ff797bcc2608d5 Mon Sep 17 00:00:00 2001
From: Vladimir Gorej <vladimir.gorej@gmail.com>
Date: Sat, 8 Dec 2018 16:37:36 +0100
Subject: [PATCH] fix(hasPath): forces hasPath to be safe

The function now doesn't throw Errors anymore when supplied with non-object.
Also fixes R.has.

Closes #2738
---
 source/hasPath.js |  7 +++++--
 test/has.js       | 10 ++++++++++
 test/hasPath.js   | 20 ++++++++++++++++++++
 3 files changed, 35 insertions(+), 2 deletions(-)

diff --git a/source/hasPath.js b/source/hasPath.js
index c3baec8..5ba0f14 100644
--- a/source/hasPath.js
+++ b/source/hasPath.js
@@ -1,6 +1,8 @@
 import _curry2 from './internal/_curry2';
 import _has from './internal/_has';
 
+import isNil from './isNil';
+
 
 /**
  * Returns whether or not a path exists in an object. Only the object's
@@ -24,13 +26,14 @@ import _has from './internal/_has';
  *      R.hasPath(['a', 'b'], {});                  // => false
  */
 var hasPath = _curry2(function hasPath(_path, obj) {
-  if (_path.length === 0) {
+  if (_path.length === 0 || isNil(obj)) {
     return false;
   }
   var val = obj;
   var idx = 0;
   while (idx < _path.length) {
-    if (_has(_path[idx], val)) {
+    console.dir(val);
+    if (!isNil(val) && _has(_path[idx], val)) {
       val = val[_path[idx]];
       idx += 1;
     } else {
diff --git a/test/has.js b/test/has.js
index f19a304..bd44498 100644
--- a/test/has.js
+++ b/test/has.js
@@ -22,4 +22,14 @@ describe('has', function() {
     eq(R.has('age', bob), false);
   });
 
+  it('tests paths on non-objects', function() {
+    eq(R.has('a', undefined), false);
+    eq(R.has('a', null), false);
+    eq(R.has('a', true), false);
+    eq(R.has('a', false), false);
+  });
+
+  it('tests currying', function() {
+    eq(R.has('a')({ a: { b: 1 } }), true);
+  });
 });
diff --git a/test/hasPath.js b/test/hasPath.js
index 97ed707..889a853 100644
--- a/test/hasPath.js
+++ b/test/hasPath.js
@@ -32,6 +32,15 @@ describe('hasPath', function() {
     eq(R.hasPath(['arrayVal', 1], obj), false);
   });
 
+  it('tests for paths in arrays', function() {
+    eq(R.hasPath([0], [1, 2]), true);
+    eq(R.hasPath([2], [1, 2]), false);
+
+    eq(R.hasPath(['0'], [1, 2]), true);
+    eq(R.hasPath(['2'], [1, 2]), false);
+  });
+
+
   it('returns false for non-existent path', function() {
     eq(R.hasPath(['Unknown'], obj), false);
     eq(R.hasPath(['objVal', 'Unknown'], obj), false);
@@ -50,4 +59,15 @@ describe('hasPath', function() {
   it('returns false for empty path', function() {
     eq(R.hasPath([], obj), false);
   });
+
+  it('tests paths on non-objects', function() {
+    eq(R.hasPath(['a', 'b'], undefined), false);
+    eq(R.hasPath(['a', 'b'], null), false);
+    eq(R.hasPath(['a', 'b'], true), false);
+    eq(R.hasPath(['a', 'b'], false), false);
+  });
+
+  it('tests currying', function() {
+    eq(R.hasPath(['a', 'b'])({ a: { b: 1 } }), true);
+  });
 });
-- 
2.17.1

```

**Comment by CrossEye** [comment]
Thanks, @char0n !
